File: python/filters/level1.py
class NpcFilter:
    def __init__(self, exclude: bool = False):
        self.exclude = exclude

    def filter(self, killmail: dict) -> bool:
        is_npc = bool(killmail.get("zkb", {}).get("npc", False))
        return not is_npc if self.exclude else is_npc


class SoloFilter:
    def __init__(self, exclude: bool = False):
        self.exclude = exclude

    def filter(self, killmail: dict) -> bool:
        is_solo = bool(killmail.get("zkb", {}).get("solo", False))
        return not is_solo if self.exclude else is_solo


class AwoxFilter:
    def __init__(self, exclude: bool = False):
        self.exclude = exclude

    def filter(self, killmail: dict) -> bool:
        is_awox = bool(killmail.get("zkb", {}).get("awox", False))
        return not is_awox if self.exclude else is_awox


class SecurityFilter:
    def __init__(self, allow: list[str]):
        self.allow = [f"loc:{z.removeprefix('loc:')}" for z in allow]

    def filter(self, killmail: dict) -> bool:
        labels = killmail.get("zkb", {}).get("labels", [])
        return any(label in self.allow for label in labels)

File: typescript/src/filters/level1.ts
import type { R2Z2Killmail } from "../types.js";
import type { KillmailFilter } from "./index.js";

export class NpcFilter implements KillmailFilter {
  constructor(private exclude: boolean = false) {}

  filter(killmail: R2Z2Killmail): boolean {
    const isNpc = killmail.zkb.npc ?? false;
    return this.exclude ? !isNpc : isNpc;
  }
}

export class SoloFilter implements KillmailFilter {
  constructor(private exclude: boolean = false) {}

  filter(killmail: R2Z2Killmail): boolean {
    const isSolo = killmail.zkb.solo ?? false;
    return this.exclude ? !isSolo : isSolo;
  }
}

export class AwoxFilter implements KillmailFilter {
  constructor(private exclude: boolean = false) {}

  filter(killmail: R2Z2Killmail): boolean {
    const isAwox = killmail.zkb.awox ?? false;
    return this.exclude ? !isAwox : isAwox;
  }
}

export class SecurityFilter implements KillmailFilter {
  private allow: string[];

  constructor(allow: string[]) {
    this.allow = allow.map((z) =>
      z.startsWith("loc:") ? z : `loc:${z}`
    );
  }

  filter(killmail: R2Z2Killmail): boolean {
    const labels = killmail.zkb.labels ?? [];
    return labels.some((label) => this.allow.includes(label));
  }
}
